### PR TITLE
Fix sequence of root cert and op cert provisioning

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
Current device commissioning flow is sending root certificate before getting the CSR from the device. As per specifications, the following should be the flow

1. Request OpCSR
2. Receive OpCSR -> Send trusted root certs
3. Receive response for root cert provisioning -> Send OpCert

#### Change overview
Updated the certificate provisioning flow to match the specifications.

#### Testing
CI runs end to end commissioning tests.
Also manually tested commissioning using Python and chip-tool controller applications.

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -1381,7 +1378,7 @@ void DeviceCommissioner::OnRootCertSuccessResponse(void * context)
 
     device = &commissioner->mActiveDevices[commissioner->mDeviceBeingPaired];
 
-    err = commissioner->SendOperationalCertificateSigningRequestCommand(device);
+    err = commissioner->ProcessOpCSR(device, device->GetCSR());
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:1381]: This doesn't make sense to me. The OpCSR should be processed by the CA before adding the root cert, as your PR description implies.

**Review by woody-apple — APPROVED** [review]

File: src/controller/CHIPDeviceController.h [hunk 2]
```diff
@@ -616,10 +616,15 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController, public SessionEst
...
+     * @param[in] CSR             The Certificate Signing Request.
+     * @param[in] CSRNonce        The Nonce sent by us when we requested the CSR.
+     * @param[in] VendorReserved1 vendor-specific information that may aid in device commissioning.
+     * @param[in] VendorReserved2 vendor-specific information that may aid in device commissioning.
+     * @param[in] VendorReserved3 vendor-specific information that may aid in device commissioning.
+     * @param[in] Signature       Cryptographic signature generated for all the above fields.
      */
-    CHIP_ERROR ProcessOpCSR(Device * device, const ByteSpan & CSR);
+    CHIP_ERROR ProcessOpCSR(const ByteSpan & CSR, const ByteSpan & CSRNonce, const ByteSpan & VendorReserved1,
+                            const ByteSpan & VendorReserved2, const ByteSpan & VendorReserved3, const ByteSpan & Signature);
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.h]: A more scalable approach is to have:

- The entire NOCSRElements as a ByteSpan
- The AttestationSignature as a ByteSpan
- A placeholder argument for 3 bytespans: DAC cert, PAI cert, PAA cert

See spec:

```
[width="99%",frame="all",options="header", cols="^,<s,<,^,^"]
|===
| Id | Field                   | Type      | Constraint                       | Conformance
| 0  | NOCSRElements           | octstr    | max <<ref_RESP_MAX,`RESP_MAX`>>  | M
| 1  | AttestationSignature    | octstr    | 64                               | M
|===
```

That would allow:

- The full payload to be passed to a CA
- The CSR portion of the `NOCSRElements` to be extracted
- Reducing the "breadth" of this public API that needs to often pass data to code that lives outside native C++ code.

The main change would be to go from "pre-processing" the inner TLV of `NOCSRElements` to post-processing it. In the synchronous call path of this command handling, it would take no additional code, just require moving the TLV decode a bit deeper in the chain.

I recommend this now because the to-spec processing of CSRResponse requires device attestation validation (which is in the works), and the public API currently makes this impossible. Since the API is being touched, it's a good time to do a little bit of reshuffling that opens the door to doing attestation signature validation on the CSR and also opening the door to end-to-end attestation of CSRReponse by CA of the commissioner (e.g. if the CA code lives in a trustzone or in a cloud, rather than in-process as is the case of the examples today).

**Review by jmartinez-silabs — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

File: src/controller/CHIPDeviceController.h (hunk 2 above)
Comment (andy31415) [reply src/controller/CHIPDeviceController.h]: @pan-apple - PRs new require all conversations to be resolved before merging. Not sure how to interpret the thumbs up: will you follow up with a commit with additional changes?

Comment (pan-apple) [reply src/controller/CHIPDeviceController.h]: Sorry, I wanted to address the review comment as part of the PR, but other PRs kept taking my time.
@tcarmelveilleux would it be ok to tackle the suggestion as a separate PR, as I cannot get back to updating this part of the code in next few days.

File: src/controller/CHIPDevice.h [hunk 3]
```diff
@@ -482,6 +492,9 @@ class DLL_EXPORT Device : public Messaging::ExchangeDelegate, public SessionEsta
 
     uint8_t mCSRNonce[kOpCSRNonceLength];
 
+    uint8_t mNOC[Credentials::kMaxDERCertLength];
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDevice.h:495]: This is invalid: where this is used, the NOC is a chain received by the CA, and it should be in TLV certificate format, and must allow for an ICAC int he chain (which this array is too small to support).

Where the call:
```
        err = commissioner->GenerateOperationalCertificates(device->GetNOC(), chipCertSpan);
        SuccessOrExit(err);
``` 

is made, looking at the logic, it sounds like:

1. Ask the device for its NOC chain
2. Generate a NOC (from a NOC?)
3. Send the NOC

There are several concepts that are mixed here and that do not match the flow's intent.

The "device" object here is used as a bucket to receive lots of temporary data from the commissioning flow, where it could instead be maintained in a commissioning context.

The flow also prevents usage of an ICAC in the chain.

I understand not all of it can be refactored at once, but here we are adding new code that does very different than intended.

At the very least we should not mix "NOC" terminology between a given commissioner's preferred internal format (which may be DER) and the actual need of the protocol: that it be TLV.

I would recommend:

1. Allocating enough space for the whole chain of NOC in matter certificate format root + ICAC + RCAC
2. When the call to the CA is made to go from NOCSR to NOC chain, keep all consideration of DER format purely internal, and output root cert and ICAC+NOC separately, in Matter format
3. Here, rename mNOC to mNOCChain, and have enough space for ICAC+NOC
4. Here, add storage for mRootCert --> Note that this is not multi-admin safe, but will do for now...
5. After obtaining the operational chain from the processing of the NOCSR, keep it "handy" while the root is being installed.

In other words, the following code:

```
    {
        MutableByteSpan chipCertSpan(chipCert.Get(), chipCertAllocatedLen);
        err = commissioner->GenerateOperationalCertificates(device->GetNOC(), chipCertSpan);
        SuccessOrExit(err);

        ChipLogProgress(Controller, "Sending operational certificate to the device");
        err = commissioner->SendOperationalCertificate(device, chipCertSpan);
        SuccessOrExit(err);
    }
```

Should only be sending the NOC (second step), rather than converting.

Comment (pan-apple) [reply src/controller/CHIPDevice.h:495]: @tcarmelveilleux, a few questions..

CA Interface (`OperationalCredentialsDelegate`)
1. The generated certificates are in DER format. I think that's the right approach, as it keeps matter specific logic (conversion to CHIPCert) contained within common part of the code. Each delegate implementation doesn't have to convert DER certs to CHIPCert before returning via the delegate API.
2. Currently, there are 3 APIs to return root cert, ICA cert and operational cert, respectively. Does it make sense to condense it to a smaller set? I think, keeping root cert API as a separate call makes sense, but ICA and opcert can definitely be combined into one.
3. Make all the APIs async. Right now only opcert API is async. 

Currently, when the commissioner first starts, it gets its own credentials using the CA Interface (`OperationalCredentialsDelegate`). The device is provisioned with the root cert that the commissioner was assigned. If that's the right thing to do, we may not need to store the root cert in the `device` objects (the `device` object is a proxy object in controller/commissioner for the actual device).

Similarly, we don't need to store the ICAC in the `device` object, as we send it to the actual device immediately after we get it from the CA. If we agree to condense the opcert, and ICAC APIs in CA Interface (`OperationalCredentialsDelegate`), it'll be required (and ok) to store both certs in `device` object.

Lastly, we currently have API (`ConvertX509CertsToChipCertArray`) to convert two x509 certs (ICAC and OpCert) to CHIPCert array. Storing the opcert temporarily in x509 format helps, as we can call `ConvertX509CertsToChipCertArray` to combine the stored opcert, and received ICAC into on CHIPCert array. If we agree to condense opcert and ICAC APIs into one API, this part of the discussion can be ignored, as we can just convert the received certs to CHIPCert and store that in `device` object.

Agree on renaming code to match the spec, and disambiguate it.

**Comment by pan-apple** [comment]
@tcarmelveilleux the latest commit tries to address your comments. Please take a look when you get a chance.
The changes are
1. Single API to request certificate chain from the CA
2. The CA is provided the NOCSRElements, attestation signatures and attestation related information.
3. The device object stores the ICAC and NOC in CHIPCert format for the duration of root certificate provisioning.

Also updated operational credentials server to generate the NOCSR in the correct format.
